### DISCUSSION TITLE
Allow to scale Texture resources

### DISCUSSION THREAD

**Discussion by JorensM** [godotengine/godot-proposals#8660] [Editor]
### Describe the project you are working on

Not a particular project, just general request

### Describe the problem or limitation you are having in your project

When an image doesn't match the desired in-game size, I either have to scale each Sprite or resize the image file. This is quite cumbersome and I'd much prefer to just be able to scale the Texture itself

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

This will allow for easier and more intuitive texture scaling. It would also be in line with the principles KISS and DRY.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

I think it should be relatively easy to add this feature

### If this enhancement will not be used often, can it be worked around with a few lines of script?

I think this would be used quite often if implemented

### Is there a reason why this should be core and not an add-on in the asset library?

I think that this would be used often enough to warrant being part of core

**Comment by paddy-exe** [comment]
This proposal doesn't really describe what you are struggling with. Godot already automatically scales images (or rather stretches them). See the image below:
<img width="1164" alt="image" src="https://github.com/godotengine/godot-proposals/assets/38077837/205e4792-ece1-4343-9364-514ad23b486e">

Furthermore, it doesn't answer the answer how this should be implemented so I will convert this here to a discussion

**Reply by JorensM** [reply]
@paddy-exe sorry I messed it up I think. I was talking about how when you use an icon in a button, the icon will take up the real size of the image, regardless of the button's size. For some reason I was under the impression that it would be the same with Sprites.

**Reply by paddy-exe** [reply]
For Icons in a button you can use [TextureButton](https://docs.godotengine.org/en/stable/classes/class_texturebutton.html#enum-texturebutton-stretchmode) which allows several StretchModes

**Reply by JorensM** [reply]
It seems that when using TextureButton you can't have the texture not take up the entire size of the button, and to have the texture alongside the text, instead of under it.

**Reply by paddy-exe** [reply]
If you want to use the Texture under the Text I would advise to create a new node for your project or compose it out of a Label and Sprite Node

**Reply by JorensM** [reply]
Ok, thanks, I guess I'll do that

**Comment by Calinou** [comment]
There's a property in Texture2D to set its internal (i.e. rendered) size independently of its original size, but it's not exposed in the inspector. I can't find the issue that refers to it right now, but it definitely exists.

**Comment by aXu-AP** [comment]
This kind of touches on a problem I had with one project. At one point I wanted to change prototype textures to higher resolution artwork. However, this resulted in all Sprite2Ds to double in size. As some sprites had animated scale, changing them was troublesome to say the least.
I think exposing Texture2D's size in importer would make sense for such cases.

**Reply by Calinou** [reply]
> I think exposing Texture2D's size in importer would make sense for such cases.

It could be exposed in the inspector for sure; my main question is whether this should be exposed as a fixed pixel size (which would allow for aspect ratio distortion) or just a scale factor like SVG.

For your particular use case, you could set the render scale to `0.5` in the import options if we exposed a scale factor.

**Reply by hsandt** [reply]
Same here, but I also had the opposite case where I switched to very small pixel art and I had to apply a factor to all my gameplay values.

I remember switching to hires art for some Unity project and I was really glad to have the texture import scale setting (esp. as _some_ art would be targetting 1080p, some 1440p, some 4K and I couldn't just change the Sprite object size itself since it would dynamically load different textures - yes not very clean but during prototyping and development, this definitely happens)

**Comment by hsandt** [comment]
I have another use case: I'm working on an HD game and since Godot measures all distances in pixels, all the distances are quite huge. So I manipulate very big gameplay parameters.

I see several solutions. First, the solutions that involve actually shrinking the scene to reduce displayed distances:
a. scale each individual sprite: cumbersome
b. scale each scene root node: feasible but I fear unintended consequences on having all the children nodes scaled
c. **(the proposal in OP)** import texture (e.g. a 4K texture) with world scale factor < 1 to that the texture appears smaller on screen, allowing me to switch the game default resolution to a lower number such as 1080p instead of 1440p or 4K. ! This should not "crunch" the texture when scale factor < 1, this would purely be an indicative number of how to _represent it in the world_ (like Unity's ppu = pixels per unit, where in this case unit = 1 px in Godot but it's not a "true pixel", just a reference) ! If player plays full screen on a 1440p or 4K monitor it will simply show everything at a higher resolution than 1080p

Then pure script solutions that don't affect the scene:
A. in my script, I manually apply some factor to all the gameplay parameters (need a lot of editing and risk of forgetting to add multiplier at some place)
B. an editor `@export` hint parameter to display values with a certain scale (e.g. all values displayed are divided by 100). Would also display a suffix not to forget we use a scale (for instance, if we divide by 1000, suffix would be "k" for "kilo"). The least risky, no change in data, only editor value display

My favorites are c (the OP) and B. In fact I'm considering opening a proposal for B.
